Standards/scs-0214-v2-k8s-node-distribution.md

```diff
@@ -0,0 +1,124 @@
+to provide a fault-tolerant and available Kubernetes cluster infrastructure.
+Since some providers only have small environments to work with and therefore couldn't
+comply with this standard, it will be treated as a RECOMMENDED standard, where providers
+can OPT OUT.
+
+If the standard is used by a provider, the following decisions are binding and valid:
+
+- The control plane nodes MUST be distributed over multiple physical machines. Kubernetes
+  provides best-practices on this topic, which are also RECOMMENDED by SCS.
+- At least one control plane instance MUST be run in each "failure zone", more are
+  RECOMMENDED in each "failure zone" to provide fault-tolerance for each zone.
+- Worker nodes are RECOMMENDED to be distributed over multiple zones. This policy makes
+  it OPTIONAL to provide a worker node in each "failure zone", meaning that worker nodes
+  can also be scaled vertically first before scaling horizontally.
+- Worker node distribution MUST be indicated to the user through some kind of labeling
+  in order to enable (anti)-affinity for workloads over "failure zones".
+- To provide metadata about the node distribution, which also enables testing of this standard,
+  providers MUST label their K8s nodes with the labels listed below.
+  - `topology.kubernetes.io/zone`
+
+    Corresponds with the label described in [K8s labels documentation][k8s-labels-docs].
+    It provides a logical zone of failure on the side of the provider, e.g. a server rack
+    in the same electrical circuit or multiple machines bound to the internet through a
+    singular network structure. How this is defined exactly is up to the plans of the provider.
+    The field gets autopopulated most of the time by either the kubelet or external mechanisms
+    like the cloud controller.
+
+  - `topology.kubernetes.io/region`
+
+    Corresponds with the label described in [K8s labels documentation][k8s-labels-docs].
+    It describes the combination of one or more failure zones into a region or domain, therefore
+    showing a larger entity of logical failure zone. An example for this could be a building
+    containing racks that are put into such a zone, since they're all prone to failure, if e.g.
+    the power for the building is cut. How this is defined exactly is also up to the provider.
+    The field gets autopopulated most of the time by either the kubelet or external mechanisms
+    like the cloud controller.
+
+  - `topology.scs.community/host-id`
+
+    This is an SCS-specific label; it MUST contain the hostID of the physical machine running
+    the hypervisor (NOT: the hostID of a virtual machine). Here, the hostID is an arbitrary identifier,
+    which doesn't need to contain the actual hostname, but it should nonetheless be unique to the host.
+    This helps identify the distribution over underlying physical machines,
+    which would be masked if VM hostIDs were used.
```

Comment:
- Probably unnecessary. I think it's clear that the hostID is a property of the physical host, so each physical host has exactly one such id. However, if we find a non-awkward way of putting it, we can still do so.
- This is beyond my depth. Maybe @matofeder can say something about this? What happens when a k8s node is migrated to another physical host?

Reply:
> Should we explicitely say something about what should happen if workload (e.g., an OpenStack VM) is migrated to another physical host? My expectation is that the host-id label must be updated.


IMO we could extend this question also with `zone` and `region` labels (AFAIK live migration is also possible across zones, and potentially some cold migration could be possible even across regions).

Currently, the approach used by KaaS V1/V2 - [CAPO](https://cluster-api-openstack.sigs.k8s.io/) reconciles 
`zone` and `region` labels, so potentially `zone` and `region` labels will be updated in case of migration. 
The `host-id` label is not covered by the current approach at all. Because of that, we introduced a custom pre-scs-standard-test task in the KaaS E2E test that does the `host-id` labeling on the fly, i.e. this `host-id` labeling workaround works also in case of hot/cold migration. 

But IMO SCS standards should be generic, and not CAPO specific, therefore I would say that the CSP SHOULD ensure that the REQUIRED labels are present and CSP SHOULD ensure that the REQUIRED labels are updated in case of hot or cold migrations. 

Currently, we do not have a production tool that ensures the above (especially for the `host-id` label), just the e2e workaround I mentioned. There is some related discussion in CAPO upstream, where we could contribute, let's see. I will keep you informed.

Reply:
What matters for me: if these migrations are relevant. And that seems to be the case. So I guess we should mention that the labels have to reflect the current state at all times (with some room for latency) or something. And yes, the reference implementation should do it ;)

Reply:
I added language that the labels must be kept up to date. Feel free to put it more elegantly. And please ignore the intermediate version where I replaced MUST by SHOULD. That was probably a bad idea.